Add hackathon page (#37)
unikraft/summer-of-code-2021: content/en/docs/hackathon.md

It is possible to solve more than one challenge and earn all points from all challenges solved.

Additional points will be awarded for:

* Signing off your commits with `git commit -s`, as part of [Developer Certificate of Origin](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) (1 point per commit) (note, teams of multiple people should have multiple sign-offs);
* Every commit should include a precise explanation of the changes made (1 point per commit);
* Every commit of the solution should leave the project in a working state (2 point per commit).

## Challenges

### 1. New `uktest` Suites, Cases & Expectations

| ID | Points | Short description |
|----|--------|-------------------|
| `uktest-entries` | 1 point per `*EXPECT*` | Use [uktest](/docs/sessions/06-testing-unikraft/#03-unikrafts-testing-framework.md) to add new tests in Unikraft components. |

#### Description

In this challenge, introduce new test suites, cases and assertions/expects in the Unikraft core repository.
New tests should be introduced as new files under a new directory `tests/` of internal microlibraries.
For example, for `vfscore`, a new directory will be located at `libs/vfscore/tests/`.

> To create a suite, for, for example, `stat()` in `vfscore`, introduce a new file `test_stat.c` where you register the suite, cases and create expectations from the `stat()` syscall provided by `vfscore`.

Reviewer: ```suggestion
To create a suite, for, to give an example, `stat()` in `vfscore`, introduce a new file `test_stat.c` where you register the suite, cases and create expectations from the `stat()` syscall provided by `vfscore`.
```

Check out the relevant POSIX document for the function in question.

Every new `*_EXPECT_*` test assertion will receive 1 point.

#### Links and Additional Resources

* [uktest](docs/sessions/06-testing-unikraft/#03-unikrafts-testing-framework.md): Unikraft's testing framework

---

### 2. Fix Build Warnings

| ID | Points | Short description |
|----|--------|-------------------|
| `fix-warnings` | 1 point per warning | Fix build warnings in Unikraft components. |

#### Description

In this challenge, fix compiler warnings during builds of internal Unikraft libraries.
(There may be other warnings in other Unikraft repositories.)

Every warning (1 line) will receive 1 point.

#### Links and Additional Resources

* [Issue on Unikraft build warnings](https://github.com/unikraft/unikraft/issues/242)

---

### 3. Add ukstore Entries

| ID | Points | Short description |
|----|--------|-------------------|
| `ukstore-entries` | 1 point per entry | Add new [ukstore](https://github.com/unikraft/unikraft/pull/275) entries. |

#### Description

> `ukstore` is a new internal library for storing and retrieving information, for example, statistics or state.

Reviewer: ```suggestion
`ukstore` is a new internal library for storing and retrieving information, such as statistics or state.
```

Introduce new counters, stats, or states via `ukstore` in other libraries that generate information.

#### Links and Additional Resources

* [ukstore](https://github.com/unikraft/unikraft/pull/275) Library

> * [Example of adding new counters to `uknetdev`](#)

Reviewer 2: TODO: Add correct link

cc. @craciunoiuc

#### Links and Additional Resources

* [Embedded Rust support in Unikraft](https://github.com/unikraft/unikraft/commit/825b1150f74a2e809341369bc19757560371b418)

---

### 15. Port PicoTCP to Unikraft

| ID | Points | Short description |
|----|--------|-------------------|
| `lib-picotcp` | 25 points | Port [PicoTCP](https://github.com/tass-belgium/picotcp) to Unikraft. |

#### Description

PicoTCP is a TCP/IP stack written in C.
Port this as an alternative to [LwIP](https://github.com/unikraft/lib-lwip) so it can be used with other network-based applications built with Unikraft.
A successful port will allow the user to replace LwIP completely with PicoTCP.

#### Links and Additional Resources

* [PicoTCP source code](https://github.com/tass-belgium/picotcp)
* [lib-lwip](https://github.com/unikraft/lib-lwip)
* [Session 04: Complex Applications](/docs/sessions/04-complex-applications/)

---

### 16. Fix an Open Bug in the Core

| ID | Points | Short description |
|----|--------|-------------------|
| `bug-fix` | Depends on bug, contact TA. | Fix an internal bug in the core. |

#### Description

There are a number of outstanding issues/bugs with the Unikraft core repository.

> To help increase the stability and solve problems for edge cases and other issues, solve an open issue which has been reported.

Reviewer 3: ```suggestion
To help increase the stability and solve problems for edge cases and other issues, solve an open issue that has been reported.
```

The top three teams will be awarded the most prestigious USoC badges:

* 3rd place: the [Green Dragon badge](https://eu.badgr.com/public/badges/WWBWO7ccTxmYLl7Om6o0Ig)
* 2nd place: the [Red Dragon badge](https://eu.badgr.com/public/badges/bTuLRsqzSKy8fZ5xrCyYSQ)
* 1st place: the [Black Dragon badge](https://eu.badgr.com/public/badges/hBNobLOGRJiydfzsibV2iw)

Challenges are listed below.
You solve a challenge:

* by submitting a pull request to the [corresponding Unikraft repository](https://github.com/unikraft/), or
* by creating a repository for a ported application / library

Points listed in challenges are a rough indication of difficulty.  It is
possible to solve more than one challenge and earn all points from all
challenges solved.

Additional points will be awarded for:

 * Signing off your commits with `git commit -s`, as part of [Developer
    Certificate of
    Origin](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) (1
    point per commit) (note, teams of multiple people should have multiple sign-offs);
 * Every commit should include a precise explanation of the changes made (1
   point per commit);

>  * Every commit of the solution should be leave the project in a working state

Reviewer 4: ```suggestion
 * Every commit of the solution should leave the project in a working state
```

   (2 point per commit);

## Challenges

### 1. New `uktest` Suites, Cases & Expectations

| ID | Points | Short description |
|----|--------|-------------------|
| `uktest-entries` | 1 point per `*EXPECT*` | Use [uktest](/docs/sessions/06-testing-unikraft/#03-unikrafts-testing-framework.md) to add new tests in Unikraft components. |

#### Description

In this challenge, introduce new test suites, cases and assertions/expects in
the Unikraft core repository.  New tests should be introduced as new files under
a new directory `tests/` of internal microlibraries.  For example, for
`vfscore`, a new directory will be located at `libs/vfscore/tests/`.

To create a suite, for, for example, `stat()` in `vfscore`, introduce a new file
`test_stat.c` where you register the suite, cases and create expectations from
the `stat()` syscall provided by `vfscore`.  Expectations and tests can usually
be checked by programming different scenarios and checking whether the return
code or `errno` is set correctly.  Check out the relevant POSIX document for

> function in question.

Reviewer 4: ```suggestion
the function in question.
```

Every new `*_EXPECT_*` test assertion will receive 1 point.

#### Links and additional resources

* [uktest](docs/sessions/06-testing-unikraft/#03-unikrafts-testing-framework.md): Unikraft's testing framework

***

### 2. Fix Build Warnings

| ID | Points | Short description |
|----|--------|-------------------|
| `fix-warnings` | 1 point per warning | Fix build warnings in Unikraft components. |

#### Description

In this challenge, fix compiler warnings during builds of internal Unikraft
libraries.

(There may be other warnings in other Unikraft repositories.)

Every warning (1 line) will receive 1 point.

#### Links and additional resources

* [Issue on Unikraft build warnings](https://github.com/unikraft/unikraft/issues/242)

***

### 3. Add ukstore Entries

| ID | Points | Short description |
|----|--------|-------------------|
| `ukstore-entries` | 1 point per entry | Add new [ukstore](https://github.com/unikraft/unikraft/pull/275) entries. |

#### Description

> `ukstore` is a new internal library for storing and retreving information, for

Reviewer 4: ```suggestion
`ukstore` is a new internal library for storing and retrieving information, for
```

example, statistics or state.  Introduce new counters, stats, or states via
`ukstore` in other libraries which generate information.

#### Links and additional resources

* [ukstore](https://github.com/unikraft/unikraft/pull/275) Library
* [Example of adding new counters to `uknetdev`](#)

***

### 3. Add new unit tests to `kraft`

| ID | Points | Short description |
|----|--------|-------------------|
| `kraft-unit-tests` | 1 point per test | Add new tests in [kraft](https://github.com/unikraft/kraft). |

#### Description

> `kraft` is a Python-based toolchain and acts as the companion tool for manging,

Reviewer 4: ```suggestion
`kraft` is a Python-based toolchain and acts as the companion tool for managing,
```

kraft list update
kraft init -t php@staging ./my-php-app
```

#### Links and additional resources

* [PHP source code](https://github.com/php/php-src)
* [PHP main website](https://www.php.net/)
* [Session 08: Basic App Porting](/docs/sessions/08-basic-app-porting/)

### 8. Port Postgres to Unikraft

| ID | Points | Short description |
|----|--------|-------------------|
| `lib-postgres` | 25 points | Port [Postgres](https://github.com/postgres/postgres) as a Unikraft library. |

#### Description

Port the object-relational database program
[Postgres](https://www.postgresql.org/) so it can run as a Unikraft unikernel.
Create a matching application component.  A user should be
able to run the project via:

```bash
kraft list add https://github.com/$USERNAME/app-postgres.git
kraft list update
kraft init -t postgres@staging ./my-postgres-app
```

#### Links and additional resources

* [Postgres source code](https://github.com/postgres/postgres)
* [Postgres main website](https://www.postgresql.org/)
* [Session 08: Basic App Porting](/docs/sessions/08-basic-app-porting/)

***

### 9. Port MySQL to Unikraft

| ID | Points | Short description |
|----|--------|-------------------|

> | `lib-mysql` | 25 points| Port [MySQl](https://github.com/mysql/mysql-server) as a Unikraft library. |

Reviewer 4: ```suggestion
| `lib-mysql` | 25 points| Port [MySQL](https://github.com/mysql/mysql-server) as a Unikraft library. |
```

#### Description

Port the relational database program [MySQL](https://www.mysql.com/) so it
can run as a Unikraft unikernel.  Create a matching application component.  A
user should be able to run the project via:

```bash
kraft list add https://github.com/$USERNAME/app-mysql.git
kraft list update
kraft init -t mysql@staging ./my-mysql-app
```

#### Links and additional resources

> * [MySQl source code](https://github.com/mysql/mysql-server)

Reviewer 4: ```suggestion
* [MySQL source code](https://github.com/mysql/mysql-server)
```

> * [MySQl main website](https://www.mysql.com/)

Reviewer 4: ```suggestion
* [MySQL main website](https://www.mysql.com/)
```

#### Description

[Tinyproxy](https://github.com/tinyproxy/tinyproxy) is a fast HTTP/HTTPS server.
Port this as a new library and application to Unikraft.  A user should be able
to run the project via:

```bash
kraft list add https://github.com/$USERNAME/app-tinyproxy.git
kraft list update
kraft init -t tinyproxy@staging ./my-tinyproxy-app
```

#### Links and additional resources

* [Tinyproxy source code](https://github.com/tinyproxy/tinyproxy)
* [Session 08: Basic App Porting](/docs/sessions/08-basic-app-porting/)

***

### 11. Port Bjoern to Unikraft

| ID | Points | Short description |
|----|--------|-------------------|
| `lib-bjoern` | 25 points | Port [Bjoern](https://github.com/jonashaag/bjoern) as a Unikraft library. |

#### Description

[Bjoern](https://github.com/jonashaag/bjoern) is a fast And ultra-lightweight
HTTP/1.1 WSGI Server.  Port this as a new library and application to Unikraft.
A user should be able to run the project via:

```bash
kraft list add https://github.com/$USERNAME/app-bjoern.git
kraft list update
kraft init -t bjoern@staging ./my-bjoern-app
```

#### Links and additional resources

> * [Bjoern source code](https://github.com/php/php-src)

Reviewer 4: ```suggestion
* [Bjoern source code](https://github.com/jonashaag/bjoern)
```

* [Session 08: Basic App Porting](/docs/sessions/08-basic-app-porting/)

***

### 12. Lua Telnet shell

| ID | Points | Short description |
|----|--------|-------------------|
| `lua-telnet` | 10 points | Make use of [lib-lua](https://github.com/unikraft/lib-lua) and add telnet server |

#### Description

[Telnet](https://en.wikipedia.org/wiki/Telnet) is a protocol for doing
bi-directional text communication.  Create a simple telnet server on top of the
Lua language and run it on Unikraft.

#### Links and additional resources

* [app-lua](https://github.com/unikraft/app-lua)
* [lib-lua](https://github.com/unikraft/lib-lua)
* [lib-lwip](https://github.com/unikraft/lib-lwip)
* [Session 04: Complex Applications](/docs/sessions/04-complex-applications/)

***

### 13. Port Memcached to Unikraft

| ID | Points | Short description |
|----|--------|-------------------|
| `lib-memcached` | 25 points | Port [memcached](https://github.com/memcached/memcached) as a Unikraft library. |

#### Description

> [memecached](https://memcached.org/) is a general purpose key-value store.  Port

Reviewer 4: ```suggestion
[memcached](https://memcached.org/) is a general purpose key-value store.  Port
```

this as a new library and application to Unikraft.  A user should be able to run
the project via:

```bash
kraft list add https://github.com/$USERNAME/app-memcached.git
kraft list update
kraft init -t memcached@staging ./my-memcached-app
```

#### Links and additional resources

* [memcached source code](https://github.com/memcached/memcached)

> * [memecached main website](https://memcached.org/)

Reviewer 4: ```suggestion
* [memcached main website](https://memcached.org/)
```
